neqo-common/src/datagram.rs

```diff
@@ -22,10 +26,43 @@ impl Datagram {
             src,
             dst,
             tos,
+            segment_size: None,
             d: d.into(),
         }
     }
+}
+
+impl<'a> Datagram<&'a [u8]> {
+    // TODO: Rename
+    pub fn new_2(src: SocketAddr, dst: SocketAddr, tos: IpTos, d: &'a [u8]) -> Self {
+        Self {
+            src,
+            dst,
+            tos,
+            segment_size: None,
+            d,
+        }
+    }
+
+    // TODO: Rename
+    pub fn new_2_with_segment_size(
```

Comment:
This bit doesn't quite match your PR comments.

Reply:
After a `recvmmsg` into a long-lived `recv_buf: Vec<u8>`, a user can instantiate a `Datagram<&'a [u8]>` via `new_2_with_segment_size`, where `d: &'a [u8]` is a view into the `recv_buf` and `segment_size: usize` is the GRO segment size returned by `recvmmsg`.

``` rust
    pub fn new_2_with_segment_size(
        src: SocketAddr,
        dst: SocketAddr,
        tos: IpTos,
        segment_size: usize,
        d: &'a [u8],
    ) -> Self {
        Self {
            src,
            dst,
            tos,
            segment_size: Some(segment_size),
            d,
        }
    }
```

https://github.com/mxinden/neqo/blob/5d847eed43007e7a68cf66fa76c9728b8a6525e4/neqo-common/src/datagram.rs#L47-L62

``` rust
// TODO: replace recv_inner in favor of this one.
pub fn recv_inner_2<'a>(
    local_address: &SocketAddr,
    state: &UdpSocketState,
    socket: impl SocketRef,
    recv_buf: &'a mut Vec<u8>,
) -> Result<Datagram<&'a [u8]>, io::Error> {

    // ...

    Ok(Datagram::new_2_with_segment_size(
        meta.addr,
        *local_address,
        meta.ecn.map(|n| IpTos::from(n as u8)).unwrap_or_default(),
        meta.stride,
        recv_buf,
    ))
}
```

https://github.com/mxinden/neqo/blob/5d847eed43007e7a68cf66fa76c9728b8a6525e4/neqo-udp/src/lib.rs#L170-L176



To avoid any confusion, the `new_2_with_segment_size` is a work-in-progress name for now.

Does this make sense?

Reply:
I significantly cleaned-up the pull request, thus hopefully avoiding the confusion above.

I am closing here. I am sorry for not making the best use of your time @martinthomson.

Please take another look whenever you can. There is no rush.